### PR TITLE
fix: handle nil request case

### LINKED ISSUE #21809: azidentity: panic on AuthenticationFailedError.Error when using custom tranporter
### Bug Report

- import path of package in question, e.g. `.../services/compute/mgmt/2018-06-01/compute`

```
github.com/Azure/azure-sdk-for-go/sdk/azidentity
```

- SDK version e.g. `master`, `latest`, `18.1.0`

`master`

  - Specify the exact commit if possible; one way to get this is the REVISION
    column output by `go list -m <module>`, for example `go list -m github.com/Azure/azure-sdk-for-go/sdk/azcore`.
- output of `go version`

```
go version go1.21.0 darwin/arm64
```

- What happened?

When using custom http transport to return failure response to `GetToken` calls, the returned `Error` calls on the returned `err` object could cause panic. This is a snippet of reproduction:

```
type unitTestTransport struct {
	h http.Handler
}

var _ http.RoundTripper = (*unitTestTransport)(nil)

func (t *unitTestTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	rec := httptest.NewRecorder()
	t.h.ServeHTTP(rec, req)

	// httptest doesn't bind Request to Response by default
	return rec.Result(), nil
}

func TestCustomTransporter(t *testing.T) {
	data, err := os.ReadFile(certPath)
	handleError(err)

	certs, key, err := azidentity.ParseCertificates(data, nil)
	handleError(err)

	testTenantID := "72f988bf-86f1-41af-91ab-2d7cd011db47"
	unitTestTransport := &unitTestTransport{
		h: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if strings.HasSuffix(r.URL.Path, "openid-configuration") {
                                 // lazy man's proxy to the first oidc discovery step. Failure from this step won't go to `AuthenticationFailedError`
				oidcConfigResp, err := http.DefaultClient.Get(r.URL.String())
				if err != nil {
					t.Errorf("unexpected error: %s", err)
				}
				defer oidcConfigResp.Body.Close()
				w.WriteHeader(oidcConfigResp.StatusCode)
				_, err = io.Copy(w, oidcConfigResp.Body)
				if err != nil {
					t.Errorf("unexpected error: %s", err)
				}
			}

			w.WriteHeader(http.StatusBadRequest) // simulate a bad request
		}),
	}
	unitTestClient := &http.Client{Transport: unitTestTransport}

	opts := &azidentity.ClientCertificateCredentialOptions{
		DisableInstanceDiscovery: true,
		ClientOptions: azcore.ClientOptions{
			Transport: unitTestClient,
		},
	}
	cred, err = azidentity.NewClientCertificateCredential(testTenantID, clientID, certs, key, opts)
	handleError(err)

	_, err = cred.GetToken(context.TODO(), policy.TokenRequestOptions{
		Scopes: []string{"test-scope"},
	})
	if err == nil {
		t.Error("unexpected nil error")
	}
	fmt.Println(err) // %!v(PANIC=Error method: runtime error: invalid memory address or nil pointer dereference)
	_ = err.Error()  // <- panic
}
```

- What did you expect or want to happen?

No panic and print out the error object.

- How can we reproduce it?

See above snippet.

- Anything we should know about your environment.

This usage is very common on unit / integration test scenarios, where we want to simulate the real world flows. Our team usually use `httptest` to create a transport object. However, according to httptest's implementation, `http.Response`'s `Request` doesn't guarantee to be presented, hence a panic will be raised when goes to this code path:

 https://github.com/Azure/azure-sdk-for-go/blob/f52645651a7e4a4ae3dacc7bca501319deeab882/sdk/azidentity/errors.go#L60

### PR DESCRIPTION
**Pull request by bcho**
Fixes #21809

- [x] The purpose of this PR is explained in this or a referenced issue.
- [x] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [ ] Updates to module CHANGELOG.md are included.
- [x] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go

### CODE REVIEW

File: sdk/azidentity/errors_test.go [hunk 1]
```diff
@@ -59,3 +59,48 @@ func TestAuthenticationFailedErrorInterface(t *testing.T) {
...
+	}
+	err := newAuthenticationFailedError(credNameAzureCLI, "error message", res, nil)
+	if e, ok := err.(*AuthenticationFailedError); ok {
+		if e.RawResponse == nil {
+			t.Fatal("expected a non-nil RawResponse")
+		}
+	} else {
+		t.Fatalf("expected AuthenticationFailedError, received %T", err)
+	}
+	errMsg := err.Error()
```
Comment (bcho) [inline sdk/azidentity/errors_test.go:90]: FYI, it looks like this:

```
AzureCLICredential authentication failed
Request information not available
--------------------------------------------------------------------------------
RESPONSE 400 Bad Request
--------------------------------------------------------------------------------
Test Error Response Body
--------------------------------------------------------------------------------
To troubleshoot, visit https://aka.ms/azsdk/go/identity/troubleshoot#azure-cli
```

File: sdk/azidentity/errors_test.go [hunk 2]
```diff
@@ -59,3 +59,48 @@ func TestAuthenticationFailedErrorInterface(t *testing.T) {
...
+	if _, ok := err.(*AuthenticationFailedError); !ok {
+		t.Fatalf("expected AuthenticationFailedError, received %T", err)
+	}
+	errMsg := err.Error()
+	if errMsg != "AzureCLICredential: error message" {
+		t.Fatalf("unexpected error message returned: %s", errMsg)
+	}
+}
+
+func TestAuthenticationFailedErrorWithoutOriginalRequest(t *testing.T) {
```
Comment (chlowell) [inline sdk/azidentity/errors_test.go:74]: Nitpicking: "original" implies we might handle differently some other category of request
```suggestion
func TestAuthenticationFailedErrorWithoutRequest(t *testing.T) {
```

File: sdk/azidentity/errors.go [hunk 3]
```diff
@@ -57,7 +57,13 @@ func (e *AuthenticationFailedError) Error() string {
 	}
 	msg := &bytes.Buffer{}
 	fmt.Fprintf(msg, e.credType+" authentication failed\n")
-	fmt.Fprintf(msg, "%s %s://%s%s\n", e.RawResponse.Request.Method, e.RawResponse.Request.URL.Scheme, e.RawResponse.Request.URL.Host, e.RawResponse.Request.URL.Path)
+	if e.RawResponse.Request != nil {
+		fmt.Fprintf(msg, "%s %s://%s%s\n", e.RawResponse.Request.Method, e.RawResponse.Request.URL.Scheme, e.RawResponse.Request.URL.Host, e.RawResponse.Request.URL.Path)
+	} else {
+		// this happens when the response is created from a custom HTTP transporter,
+		// which doesn't guarantee to bind the original request to the response
+		fmt.Fprint(msg, "Request information not available\n")
```
Comment (chlowell) [inline sdk/azidentity/errors.go:65]:
```suggestion
		fmt.Fprintln(msg, "Request information not available")
```

**Review by chlowell — APPROVED** [review]
Thanks for doing this!

**Comment by bcho** [comment]
@chlowell I have merged with your suggestions, PTAL!

**Review by chlowell — APPROVED** [review]

**Pull request merged**